Hello Johannah!

My name is Dmitry.  I'm glad to review your work today.
I will mark your mistakes and give you some hints how it is possible to fix them. We are getting ready for real job, where your team leader/senior colleague will do exactly the same. Don't worry and study with pleasure! 

Below you will find my comments - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Success. Everything is done succesfully.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Remarks. Some recommendations.
</div>

<div class="alert alert-block alert-danger">

<b>Reviewer's comment</b> <a class="tocSkip"></a>

Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>

You can answer me by using this:

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

Text here.
</div>

# Analyzing borrowers’ risk of defaulting

My project is to prepare a report for a bank’s loan division. I’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

My report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.

My hypothesis is that there is a coorelation between a borrowers chance of defaulting on a loan when looking at their income, education and reason for requesting the loan. I will test my hypothesis in the following analysis 



<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Great start with an introduction!
</div>

## Open the data file and have a look at the general information. 


In [1]:
import pandas as pd


df= pd.read_csv('/datasets/credit_scoring_eng.csv')


## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

Now let's explore the data. I want to see how many columns and rows it has, and look at a few rows to check for potential issues.

In [2]:

df.describe()


children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [3]:

df.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family


I am seeing that the education column has mixed upper and lower case useage, to be more uniform I will change that column to all lower case. I will also convert the float data type in the 'days_employed' ans 'total_income' columns to an integer type. The 'days_employed' column shows many negative values. I am not sure what that means and will need to investigate further.

In [4]:
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB



The columns 'days_employed' and 'total_income are missing some data. The rest of the columns appear to be complete.

In [5]:

print(df['days_employed'].isna().sum())
print(df['total_income'].isna().sum())

2174
2174


 The missing data in the two columns appears to be symmetrical. In order to check this I will check all the rows with missing data to confirm the numbers are actually related. The code below indicates the missing values in the columns are also equal to the missing columns in the rows. The rows that are missing data for 'days_employed' are also missing data for 'total_income'

In [6]:

df.isna().sum(axis='rows')


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [7]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
filtered_df = df.loc[(df['days_employed'].isna())&(df['total_income'].isna())]

print(filtered_df.count())
print(filtered_df.head(5))


children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64
    children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29         

**Intermediate conclusion**

The number of rows in the filtered table matches the number of missing values. We can conclude that the missing data is missing from the same form in each entry. To determine if the missing values account for a significant piece of the data I will find the percentage of the missing values compared to the total dataset.



[Explain your next steps and how they correlate with the conclusions you made so far.]

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correct.
</div>

In [8]:
df.isna().sum()*100/len(df)

children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

The missing data accounts for around 10% of the total dataset. This is significant, I will be investigating why the values are missing for these entries. I suspect employment status such as retirement and unemployment could be a reason why days employed and income would be empty.In my next steps I will look into these reasonings 

In [9]:

filtered_df[filtered_df['income_type']=='retiree']



children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
29            0            NaN         63  secondary education             1   
55            0            NaN         54  secondary education             1   
67            0            NaN         52    bachelor's degree             0   
145           0            NaN         62  secondary education             1   
...         ...            ...        ...                  ...           ...   
21311         0            NaN         49  secondary education             1   
21321         0            NaN         56  Secondary Education             1   
21414         0            NaN         65  secondary education             1   
21415         0            NaN         54  secondary education             1   
21423         0            NaN         63  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
12     civil partnership                 1      M     retiree     0   
29             unmarried                 4      F     retiree     0   
55     civil partnership                 1      F     retiree     1   
67               married                 0      F     retiree     0   
145              married                 0      M     retiree     0   
...                  ...               ...    ...         ...   ...   
21311            married                 0      F     retiree     0   
21321            married                 0      F     retiree     0   
21414            married                 0      F     retiree     0   
21415            married                 0      F     retiree     0   
21423            married                 0      M     retiree     0   

       total_income                              purpose  
12              NaN                    to have a wedding  
29              NaN               building a real estate  
55              NaN                    to have a wedding  
67              NaN  purchase of the house for my family  
145             NaN                  building a property  
...             ...                                  ...  
21311           NaN      buying property for renting out  
21321           NaN             real estate transactions  
21414           NaN             purchase of my own house  
21415           NaN                 housing transactions  
21423           NaN                    purchase of a car  

[413 rows x 12 columns]

In [10]:
filtered_df[filtered_df['days_employed'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

Based on these filters I see that there is more detail to these missing values. Some seem to be random but others could be due to retirement or unemployment.This is definitly a blind spot in this data set ( that there is no option for reporting unemployment, or there is and people dont want to report it and seem unworthy of credit).  It may be better to fill with missing values with categorized medians/means. This way the blank incomes are standardized and more accurate than leaving them blank, especially for further analysis.   


**Possible reasons for missing values in data**

I think the values may be missing due to random error, unemployment, and retirement. I have already seen some pattern showing retirement could be a reason for the missing data in the filtered tables bit now I will check the whole data set to see the distribution from a different angle.


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

You are right!
</div>

In [11]:

df.loc[df['income_type']=='retiree']

children  days_employed  dob_years            education  education_id  \
4             0  340266.072047         53  secondary education             1   
12            0            NaN         65  secondary education             1   
18            0  400281.136913         53  secondary education             1   
24            1  338551.952911         57  secondary education             1   
25            0  363548.489348         67  secondary education             1   
...         ...            ...        ...                  ...           ...   
21505         0  338904.866406         53  secondary education             1   
21508         0  386497.714078         62  secondary education             1   
21509         0  362161.054124         59    bachelor's degree             0   
21518         0  373995.710838         59  SECONDARY EDUCATION             1   
21521         0  343937.404131         67  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
4      civil partnership                 1      F     retiree     0   
12     civil partnership                 1      M     retiree     0   
18       widow / widower                 2      F     retiree     0   
24             unmarried                 4      F     retiree     0   
25               married                 0      M     retiree     0   
...                  ...               ...    ...         ...   ...   
21505  civil partnership                 1      M     retiree     0   
21508            married                 0      M     retiree     0   
21509            married                 0      M     retiree     0   
21518            married                 0      F     retiree     0   
21521            married                 0      F     retiree     0   

       total_income                                   purpose  
4         25378.572                         to have a wedding  
12              NaN                         to have a wedding  
18         9091.804                  buying a second-hand car  
24        46487.558  transactions with commercial real estate  
25         8818.041                           buy real estate  
...             ...                                       ...  
21505     12070.399                         to have a wedding  
21508     11622.175                                  property  
21509     11684.650                  real estate transactions  
21518     24618.344                         purchase of a car  
21521     24959.969                         purchase of a car  

[3856 rows x 12 columns]

In [12]:
df.loc[df['days_employed'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Intermediate conclusion**

Interesting! The distribution is not similar to that in the filtered table. This makes me think it could be due to random error.


**Conclusions**

After looking into filtered tables and the wider dataset I did not find a patern in the distribution of the missing data. This leads me to believe it is random.
Because the data is missing in total_income and days_employed I will need to consider the categories I can use to help fill the data and whether to use means or medians. Next I will transform the data by fixing the duplicates, redundant entries, incorrect artifacts, and missing values.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Good job.
</div>

## Data transformation

Ill go through each column to see where there are issues. 


In [13]:

df['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64

In [14]:
df['education']= df['education'].str.lower()

In [15]:

df['education'].value_counts()

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

By lowering the case I was able to consolidate the values in the education column. Nice!

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Great!
</div>

In [16]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

 There are negative values in this column which must be accidental. I am also skeptical that there are 76 people with 20 children. I believe this is an accident as well. 

In [17]:
children_percentage= [(76+47)*100/df['children'].sum()]
print(children_percentage)

[1.0603448275862069]


In [18]:
df=df[df.children !=-1]
df=df[df.children !=20]

Because these values are making up only one percent of the data I think it is best to drop the rows where people reported having -1 or 20 children 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

We can also say -1 = 1 and 20 = 2
</div>

In [19]:
df['children'].value_counts()


0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

I will check the 'days_employed' column because I noticed in the head() that there weres some negative values reported

In [20]:

df['days_employed'].value_counts()

-327.685916     1
-1580.622577    1
-4122.460569    1
-2828.237691    1
-2636.090517    1
               ..
-201.643573     1
-7120.517564    1
-2146.884040    1
-881.454684     1
-3382.113891    1
Name: days_employed, Length: 19240, dtype: int64

In [21]:
negative_days= df[df['days_employed']<0].count()
print(negative_days*100/len(df['days_employed']))


children            73.866928
days_employed       73.866928
dob_years           73.866928
education           73.866928
education_id        73.866928
family_status       73.866928
family_status_id    73.866928
gender              73.866928
income_type         73.866928
debt                73.866928
total_income        73.866928
purpose             73.866928
dtype: float64


73% of the entries are negative in 'days_employed' this must be a technical issue. Perhaps they are using days passed? I will convert the negative days to positive. 

In [22]:
df['days_employed']= df['days_employed'].abs()

In [23]:

df['days_employed'].value_counts()

142.276217     1
402.974768     1
3601.450735    1
1849.622944    1
5849.845620    1
              ..
3951.652030    1
847.043824     1
1745.884477    1
1801.512744    1
1636.419775    1
Name: days_employed, Length: 19240, dtype: int64

In [24]:

df.loc[df['days_employed']<0]


Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Well done!
</div>

In [25]:


df['dob_years'].value_counts()

35    614
40    603
41    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
29    543
44    543
48    536
30    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
53    457
57    457
51    446
59    441
55    441
26    406
60    376
25    356
61    353
62    351
63    268
24    263
64    263
23    252
65    194
66    183
22    183
67    167
21    110
0     100
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

There are 101 people reporting their age as 0.They account for .4% of the data. I will drop those rows.

In [26]:
df[df['dob_years']==0].count()*100/len(df['dob_years'])

children            0.467246
days_employed       0.420521
dob_years           0.467246
education           0.467246
education_id        0.467246
family_status       0.467246
family_status_id    0.467246
gender              0.467246
income_type         0.467246
debt                0.467246
total_income        0.420521
purpose             0.467246
dtype: float64

The 0 age makes up .4% of the data, I am comfortable dropping those rows.

In [27]:
df=df[df.dob_years !=0]

In [28]:
df['dob_years'].value_counts()

35    614
41    603
40    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
44    543
29    543
48    536
30    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
57    457
53    457
51    446
55    441
59    441
26    406
60    376
25    356
61    353
62    351
63    268
64    263
24    263
23    252
65    194
22    183
66    183
67    167
21    110
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [29]:

df['family_status'].value_counts()

married              12254
civil partnership     4139
unmarried             2783
divorced              1179
widow / widower        947
Name: family_status, dtype: int64

No problems here

In [30]:

df['gender'].value_counts()

F      14083
M       7218
XNA        1
Name: gender, dtype: int64

In [31]:
df=df[df.gender !='XNA']

In [32]:
df['gender'].value_counts()

F    14083
M     7218
Name: gender, dtype: int64

I dropped XNA as it appears to be a technical error.



<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

<s>"XNA" seems to be technical error. Better to drop it.
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Well done!
</div>

In [33]:

df['income_type'].value_counts()

employee                       10996
business                        5033
retiree                         3819
civil servant                   1447
unemployed                         2
entrepreneur                       2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

All good

Now I will check for duplicates

In [34]:


df.duplicated().sum()

71

In [35]:
df_duplicated = df[df.duplicated()]
df_duplicated

children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
3290          0            NaN         58  secondary education             1   
4182          1            NaN         34    bachelor's degree             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0            NaN         64  secondary education             1   
21032         0            NaN         60  secondary education             1   
21132         0            NaN         47  secondary education             1   
21281         1            NaN         30    bachelor's degree             0   
21415         0            NaN         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849            NaN  purchase of the house for my family  
3290            NaN                    to have a wedding  
4182            NaN                     wedding ceremony  
4851            NaN                     wedding ceremony  
5557            NaN                    to have a wedding  
...             ...                                  ...  
20702           NaN              supplementary education  
21032           NaN                   to become educated  
21132           NaN                   housing renovation  
21281           NaN           buy commercial real estate  
21415           NaN                 housing transactions  

[71 rows x 12 columns]

In [36]:
df.duplicated().sum()*100/len(df)

0.33331768461574574

making up only .33% I will drop these duplicates.

In [37]:

df = df.drop_duplicates().reset_index()

In [38]:

df.duplicated().sum()

0

In [39]:

len(df)

21230

In [40]:
len(df)/21525 #this is the length of the original dataset

0.986295005807201



The dataframe has been cleaned up and I was able to conserve 98.6% of the data. Duplicated rows were dropped. All columns were checked for strange values, if any were found they were either dropped or fixed. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Great job!
</div>

# Working with missing values

Based on the IDs provided I know that I can make a couple dictionaries based on family status:family_status_id and 
education:education_id


In [41]:
family_dict= df.set_index('family_status_id').to_dict()['family_status']
print(family_dict)

{0: 'married', 1: 'civil partnership', 2: 'widow / widower', 3: 'divorced', 4: 'unmarried'}


In [42]:
education_dict=df.set_index('education_id').to_dict()['education']
print(education_dict)

{0: "bachelor's degree", 1: 'secondary education', 2: 'some college', 3: 'primary education', 4: 'graduate degree'}


### Restoring missing values in `total_income`


As I found earlier there are missing values in total_income and in days_employed. I will fix them by filling them with an appropriate mean or median based on my investigation below

To help me find a good fit for the population I will create a category for the ages of the clients. This will be a new column in the dataframe.


In [43]:
def age_group(dob_years):
    
    if dob_years <= 30 :
        return 'young adult'
    if dob_years <= 50 :
        return 'adult'
    return 'older adult'
    

In [44]:
# Test if the function works
print(age_group(28))
print(age_group(48))
print(age_group(58))


young adult
adult
older adult


In [45]:

df['age_group'] = df['dob_years'].apply(age_group)
df.head(20)


index  children  days_employed  dob_years            education  \
0       0         1    8437.673028         42    bachelor's degree   
1       1         1    4024.803754         36  secondary education   
2       2         0    5623.422610         33  secondary education   
3       3         3    4124.747207         32  secondary education   
4       4         0  340266.072047         53  secondary education   
5       5         0     926.185831         27    bachelor's degree   
6       6         0    2879.202052         43    bachelor's degree   
7       7         0     152.779569         50  secondary education   
8       8         2    6929.865299         35    bachelor's degree   
9       9         0    2188.756445         41  secondary education   
10     10         2    4171.483647         36    bachelor's degree   
11     11         0     792.701887         40  secondary education   
12     12         0            NaN         65  secondary education   
13     13         0    1846.641941         54         some college   
14     14         0    1844.956182         56    bachelor's degree   
15     15         1     972.364419         26  secondary education   
16     16         0    1719.934226         35  secondary education   
17     17         0    2369.999720         33    bachelor's degree   
18     18         0  400281.136913         53  secondary education   
19     19         0   10038.818549         48  secondary education   

    education_id      family_status  family_status_id gender income_type  \
0              0            married                 0      F    employee   
1              1            married                 0      F    employee   
2              1            married                 0      M    employee   
3              1            married                 0      M    employee   
4              1  civil partnership                 1      F     retiree   
5              0  civil partnership                 1      M    business   
6              0            married                 0      F    business   
7              1            married                 0      M    employee   
8              0  civil partnership                 1      F    employee   
9              1            married                 0      M    employee   
10             0            married                 0      M    business   
11             1            married                 0      F    employee   
12             1  civil partnership                 1      M     retiree   
13             2            married                 0      F    employee   
14             0  civil partnership                 1      F    business   
15             1            married                 0      F    employee   
16             1            married                 0      F    employee   
17             0  civil partnership                 1      M    employee   
18             1    widow / widower                 2      F     retiree   
19             1           divorced                 3      F    employee   

    debt  total_income                              purpose    age_group  
0      0     40620.102                purchase of the house        adult  
1      0     17932.802                         car purchase        adult  
2      0     23341.752                purchase of the house        adult  
3      0     42820.568              supplementary education        adult  
4      0     25378.572                    to have a wedding  older adult  
5      0     40922.170                purchase of the house  young adult  
6      0     38484.156                 housing transactions        adult  
7      0     21731.829                            education        adult  
8      0     15337.093                     having a wedding        adult  
9      0     23108.150  purchase of the house for my family        adult  
10     0     18230.959                      buy real estate        adult  
11     0     12331.077           buy

In [46]:
df['age_group'].value_counts()


adult          10918
older adult     6614
young adult     3698
Name: age_group, dtype: int64

 Thinking about income: It typically depends on ones employment, which could depend on ones education, as well as ones age. I am going to look at the distribution of these factors and how they effect ones income in these data. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Nice approach!
</div>

In [47]:

df_drop= df.dropna()
df_drop.info()
df_drop.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19149 entries, 0 to 21229
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             19149 non-null  int64  
 1   children          19149 non-null  int64  
 2   days_employed     19149 non-null  float64
 3   dob_years         19149 non-null  int64  
 4   education         19149 non-null  object 
 5   education_id      19149 non-null  int64  
 6   family_status     19149 non-null  object 
 7   family_status_id  19149 non-null  int64  
 8   gender            19149 non-null  object 
 9   income_type       19149 non-null  object 
 10  debt              19149 non-null  int64  
 11  total_income      19149 non-null  float64
 12  purpose           19149 non-null  object 
 13  age_group         19149 non-null  object 
dtypes: float64(2), int64(6), object(6)
memory usage: 2.2+ MB


index  children  days_employed  dob_years            education  \
0      0         1    8437.673028         42    bachelor's degree   
1      1         1    4024.803754         36  secondary education   
2      2         0    5623.422610         33  secondary education   
3      3         3    4124.747207         32  secondary education   
4      4         0  340266.072047         53  secondary education   

   education_id      family_status  family_status_id gender income_type  debt  \
0             0            married                 0      F    employee     0   
1             1            married                 0      F    employee     0   
2             1            married                 0      M    employee     0   
3             1            married                 0      M    employee     0   
4             1  civil partnership                 1      F     retiree     0   

   total_income                  purpose    age_group  
0     40620.102    purchase of the house        adult  
1     17932.802             car purchase        adult  
2     23341.752    purchase of the house        adult  
3     42820.568  supplementary education        adult  
4     25378.572        to have a wedding  older adult

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

<s>Please check comment in the end of block before filling NANs.
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

<s>Let's take a look at how to improve your approach. <br>
    Income depends on several factors. Consider to pivot data with several levels to find suitable filler for NAN in every group. <br>
    Please check https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pivot.html
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

<s>Let's go deeper in working with missing values in this task. <br>
    Our recommendation here is to group people by several characteristics for more accurate result in median/mean. <br>
    Propably you will need custom function to fill NANs with result from pivot table =) <br>
You've already done a lot, I am sure you'll take care of it! <br>
    Please check example below:
</div>

In [56]:
# reviewer's example
#df.pivot_table(index=['age_group','income_type' ], columns='education', values='total_income', aggfunc='median')

In [97]:
table = df_drop.pivot_table(index=['age_group','income_type' ], columns='education', values='total_income', aggfunc='median')
table

education                                bachelor's degree  graduate degree  \
age_group   income_type                                                       
adult       business                            33702.3130              NaN   
            civil servant                       28178.5190       17822.7570   
            employee                            27096.0190       25161.5835   
            paternity / maternity leave                NaN              NaN   
            retiree                             32491.0720              NaN   
            unemployed                          32435.6020              NaN   
older adult business                            34987.9720              NaN   
            civil servant                       29314.0030              NaN   
            employee                            27979.3760       42945.7940   
            retiree                             22784.6850       28334.2150   
young adult business                            28448.9855              NaN   
            civil servant                       24483.1205              NaN   
            employee                            24434.1270              NaN   
            entrepreneur                        79866.1030              NaN   
            retiree                             34394.1280              NaN   
            student                             15712.2600              NaN   

education                                primary education  \
age_group   income_type                                      
adult       business                            27305.9010   
            civil servant                       30545.9490   
            employee                            19887.1210   
            paternity / maternity leave                NaN   
            retiree                             15412.8750   
            unemployed                                 NaN   
older adult business                            23066.1310   
            civil servant                       12709.2755   
            employee                            20896.0125   
            retiree                             16415.7850   
young adult business                            21167.1430   
            civil servant                       30563.3830   
            employee                            23533.5420   
            entrepreneur                               NaN   
            retiree                                    NaN   
            student                                    NaN   

education                                secondary education  some college  
age_group   income_type                                                     
adult       business                              26144.0300    31437.1180  
            civil servant                         21740.9130    29281.5325  
            employee                              22166.6270    27035.7210  
            paternity / maternity leave            8612.6610           NaN  
            retiree                               21219.0680     7015.4225  
            unemployed                             9593.1190           NaN  
older adult business                              25063.4520    33073.7630  
            civil servant                         22940.3780           NaN  
            employee                              21854.6310    27143.4650  
            retiree                               18259.1760    19633.0750  
young adult business                              23912.5945    25570.8550  
            civil servant                         21032.8640    21640.4795  
            employee                              20273.9605    22024.3690  
            entrepreneur                                 NaN           NaN  
            retiree                               12807.0710           NaN  
            student                                      NaN           NaN

If I group the data by age_group, education level, and income type and create medians for individuals based on these groupings I believe this will be the most concise and accurate way to fill the missing incomes

In [58]:
median = df.groupby(['age_group', 'education', 'income_type'])['total_income'].transform('median')

df['total_income'] = df['total_income'].fillna(median)

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

This is a good approach too. Well done!
</div>

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             21230 non-null  int64  
 1   children          21230 non-null  int64  
 2   days_employed     19149 non-null  float64
 3   dob_years         21230 non-null  int64  
 4   education         21230 non-null  object 
 5   education_id      21230 non-null  int64  
 6   family_status     21230 non-null  object 
 7   family_status_id  21230 non-null  int64  
 8   gender            21230 non-null  object 
 9   income_type       21230 non-null  object 
 10  debt              21230 non-null  int64  
 11  total_income      21229 non-null  float64
 12  purpose           21230 non-null  object 
 13  age_group         21230 non-null  object 
dtypes: float64(2), int64(6), object(6)
memory usage: 2.3+ MB


I see there is one NaN remaining. Lets look closer. 

In [60]:
df[df['total_income'].isna()]

index  children  days_employed  dob_years          education  \
5880   5936         0            NaN         58  bachelor's degree   

      education_id family_status  family_status_id gender   income_type  debt  \
5880             0       married                 0      M  entrepreneur     0   

      total_income                      purpose    age_group  
5880           NaN  buy residential real estate  older adult

This person is quite unique and a median for their circumstance did not make it into my table. I will fill their income based on their income_type and education.

In [61]:
df.loc[5880,['total_income']]= 79866.1030


In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             21230 non-null  int64  
 1   children          21230 non-null  int64  
 2   days_employed     19149 non-null  float64
 3   dob_years         21230 non-null  int64  
 4   education         21230 non-null  object 
 5   education_id      21230 non-null  int64  
 6   family_status     21230 non-null  object 
 7   family_status_id  21230 non-null  int64  
 8   gender            21230 non-null  object 
 9   income_type       21230 non-null  object 
 10  debt              21230 non-null  int64  
 11  total_income      21230 non-null  float64
 12  purpose           21230 non-null  object 
 13  age_group         21230 non-null  object 
dtypes: float64(2), int64(6), object(6)
memory usage: 2.3+ MB


In [94]:
df['total_income'].value_counts()

22166    480
18259    294
26144    187
20273    162
21854    146
        ... 
12160      1
14211      1
6023       1
18317      1
52973      1
Name: total_income, Length: 15276, dtype: int64

In [95]:
df['total_income']=df.total_income.astype(int)
df['total_income'].head()
df.head(10)

index  children  days_employed  dob_years            education  \
0      0         1           8437         42    bachelor's degree   
1      1         1           4024         36  secondary education   
2      2         0           5623         33  secondary education   
3      3         3           4124         32  secondary education   
4      4         0         340266         53  secondary education   
5      5         0            926         27    bachelor's degree   
6      6         0           2879         43    bachelor's degree   
7      7         0            152         50  secondary education   
8      8         2           6929         35    bachelor's degree   
9      9         0           2188         41  secondary education   

   education_id      family_status  family_status_id gender income_type  debt  \
0             0            married                 0      F    employee     0   
1             1            married                 0      F    employee     0   
2             1            married                 0      M    employee     0   
3             1            married                 0      M    employee     0   
4             1  civil partnership                 1      F     retiree     0   
5             0  civil partnership                 1      M    business     0   
6             0            married                 0      F    business     0   
7             1            married                 0      M    employee     0   
8             0  civil partnership                 1      F    employee     0   
9             1            married                 0      M    employee     0   

   total_income                              purpose    age_group  \
0         40620                purchase of the house        adult   
1         17932                         car purchase        adult   
2         23341                purchase of the house        adult   
3         42820              supplementary education        adult   
4         25378                    to have a wedding  older adult   
5         40922                purchase of the house  young adult   
6         38484                 housing transactions        adult   
7         21731                            education        adult   
8         15337                     having a wedding        adult   
9         23108  purchase of the house for my family        adult   

  purpose_category   income_level  
0          housing  middle income  
1  car transaction     low income  
2          housing  middle income  
3        education  middle income  
4          wedding  middle income  
5          housing  middle income  
6          housing  middle income  
7        education  middle income  
8          wedding     low income  
9          housing  middle income

All the missing total_income values have been replaced with the median for each entries age group, income type, and education level. I have also changed the data type to integer. 

###  Restoring values in `days_employed`

 I am going to do a similar analysis to what I did in for the other missing values. This will help me determine the best fit for the population. Thinking about what may impact ones days employed, age is a big factor as well as education level.

In [64]:
table = pd.pivot_table(data = df_drop,
                      index = ['age_group'],
                       values = 'days_employed',
                       aggfunc = ['mean', 'median']
                      )
table

mean         median
             days_employed  days_employed
age_group                                
adult         10395.136690    1867.970299
older adult  197541.560600  333946.445320
young adult    1925.799317    1046.095163

In [65]:
table = pd.pivot_table(data = df_drop,
                      index = ['gender'],
                       values = 'days_employed',
                       aggfunc = ['mean', 'median']
                      )
table

mean        median
       days_employed days_employed
gender                            
F       82474.919277   2540.740460
M       37176.432951   1667.140839

In [66]:
table = pd.pivot_table(data = df_drop,
                      index = ['education'],
                       values = 'days_employed',
                       aggfunc = ['mean', 'median']
                      )
table

mean        median
                     days_employed days_employed
education                                       
bachelor's degree     42451.577484   1896.569279
graduate degree      121323.630206   5660.057032
primary education    130340.426349   3043.933615
secondary education   76508.592640   2393.978448
some college          20823.242105   1209.230373

In [67]:
table = pd.pivot_table(data = df_drop,
                      index = ['income_type'],
                       values = 'days_employed',
                       aggfunc = ['mean', 'median']
                      )
table

mean         median
                             days_employed  days_employed
income_type                                              
business                       2119.512866    1555.993659
civil servant                  3392.119263    2672.903939
employee                       2325.740892    1573.791064
entrepreneur                    520.848083     520.848083
paternity / maternity leave    3296.759962    3296.759962
retiree                      365037.486121  365269.100414
student                         578.751554     578.751554
unemployed                   366413.652744  366413.652744

I am going to use the medians for the income type to fill the days employed. What a person reports as their job is a good indicator of how long a person would be employed. Also the distribution of the age group was pretty extreme, I would not feel comfortable filling unknow values with those numbers. Filling with the median helps remove any issues with outliers skewing our distribution. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correct.
</div>

In [68]:
df['days_employed'] = (df.groupby('income_type')['days_employed']
                      .transform(lambda x: x.fillna(x.median())))


In [69]:
df['days_employed']=df['days_employed'].astype('int')

Now that I have resolved the NaN values I am able to convert the data type from float to integer.

In [70]:

df['days_employed'].value_counts()

1573      1067
1555       497
365269     380
2672       146
327         16
          ... 
360640       1
391558       1
395796       1
4475         1
2049         1
Name: days_employed, Length: 9041, dtype: int64

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             21230 non-null  int64  
 1   children          21230 non-null  int64  
 2   days_employed     21230 non-null  int64  
 3   dob_years         21230 non-null  int64  
 4   education         21230 non-null  object 
 5   education_id      21230 non-null  int64  
 6   family_status     21230 non-null  object 
 7   family_status_id  21230 non-null  int64  
 8   gender            21230 non-null  object 
 9   income_type       21230 non-null  object 
 10  debt              21230 non-null  int64  
 11  total_income      21230 non-null  float64
 12  purpose           21230 non-null  object 
 13  age_group         21230 non-null  object 
dtypes: float64(1), int64(7), object(6)
memory usage: 2.3+ MB


All the missing values have been filled. 

## Categorization of data

 To help me complete my task I will categorize the data based off of the categories of interest to the bank. These are: debt, children, family status, and purpose for the loan.

In [72]:
df_cat= df[['family_status','children','debt','purpose']]

df_cat.head(15)

family_status  children  debt                              purpose
0             married         1     0                purchase of the house
1             married         1     0                         car purchase
2             married         0     0                purchase of the house
3             married         3     0              supplementary education
4   civil partnership         0     0                    to have a wedding
5   civil partnership         0     0                purchase of the house
6             married         0     0                 housing transactions
7             married         0     0                            education
8   civil partnership         2     0                     having a wedding
9             married         0     0  purchase of the house for my family
10            married         2     0                      buy real estate
11            married         0     0           buy commercial real estate
12  civil partnership         0     0                    to have a wedding
13            married         0     0                         car purchase
14  civil partnership         0     1          buy residential real estate

I can see there is a need for a few category in the purpose column. I will find the unique values to see which categories I will make.

In [73]:
df_cat.purpose.unique()


array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

 I can see I need a category for real estate, housing, education, car purchase, and  wedding. lets check the rest of the columns before moving on. 

In [74]:
df_cat.family_status.unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

In [75]:
df_cat.debt.unique()

array([0, 1])

In [76]:
df_cat.children.unique()

array([1, 0, 3, 2, 4, 5])

I am seeing that I can make categories for the purposes for the loan requests. I will make a new column for these catagories. I will make a category for education, wedding, car, real estate, house

In [77]:
df['purpose'].value_counts()

wedding ceremony                            785
having a wedding                            759
to have a wedding                           755
real estate transactions                    669
buy commercial real estate                  655
buying property for renting out             647
transactions with commercial real estate    643
housing transactions                        641
purchase of the house for my family         636
housing                                     635
purchase of the house                       634
property                                    627
construction of own property                626
transactions with my real estate            623
building a property                         619
purchase of my own house                    618
building a real estate                      617
buy real estate                             612
housing renovation                          602
buy residential real estate                 599
buying my own car                       

In [78]:
def purpose_category(purpose):
    if 'wedding'in purpose:
        return 'wedding'
    if 'real estate'in purpose:
        return 'real estate'
    if 'car' in purpose:
        return 'car transaction'
    if 'education'in purpose:
        return 'education'
    if 'property' in purpose:
        return 'real estate'
    if 'house' or 'housing' in purpose:
        return 'housing'

In [79]:
df['purpose_category']= df['purpose'].apply(purpose_category)


In [80]:
df['purpose_category'].count()

21230

In [81]:
df['purpose_category'].value_counts()

real estate        6937
housing            4666
car transaction    4258
education          3070
wedding            2299
Name: purpose_category, dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Well done!
</div>

In [82]:
df.head(20)

index  children  days_employed  dob_years            education  \
0       0         1           8437         42    bachelor's degree   
1       1         1           4024         36  secondary education   
2       2         0           5623         33  secondary education   
3       3         3           4124         32  secondary education   
4       4         0         340266         53  secondary education   
5       5         0            926         27    bachelor's degree   
6       6         0           2879         43    bachelor's degree   
7       7         0            152         50  secondary education   
8       8         2           6929         35    bachelor's degree   
9       9         0           2188         41  secondary education   
10     10         2           4171         36    bachelor's degree   
11     11         0            792         40  secondary education   
12     12         0         365269         65  secondary education   
13     13         0           1846         54         some college   
14     14         0           1844         56    bachelor's degree   
15     15         1            972         26  secondary education   
16     16         0           1719         35  secondary education   
17     17         0           2369         33    bachelor's degree   
18     18         0         400281         53  secondary education   
19     19         0          10038         48  secondary education   

    education_id      family_status  family_status_id gender income_type  \
0              0            married                 0      F    employee   
1              1            married                 0      F    employee   
2              1            married                 0      M    employee   
3              1            married                 0      M    employee   
4              1  civil partnership                 1      F     retiree   
5              0  civil partnership                 1      M    business   
6              0            married                 0      F    business   
7              1            married                 0      M    employee   
8              0  civil partnership                 1      F    employee   
9              1            married                 0      M    employee   
10             0            married                 0      M    business   
11             1            married                 0      F    employee   
12             1  civil partnership                 1      M     retiree   
13             2            married                 0      F    employee   
14             0  civil partnership                 1      F    business   
15             1            married                 0      F    employee   
16             1            married                 0      F    employee   
17             0  civil partnership                 1      M    employee   
18             1    widow / widower                 2      F     retiree   
19             1           divorced                 3      F    employee   

    debt  total_income                              purpose    age_group  \
0      0     40620.102                purchase of the house        adult   
1      0     17932.802                         car purchase        adult   
2      0     23341.752                purchase of the house        adult   
3      0     42820.568              supplementary education        adult   
4      0     25378.572                    to have a wedding  older adult   
5      0     40922.170                purchase of the house  young adult   
6      0     38484.156                 housing transactions        adult   
7      0     21731.829                            education        adult   
8      0     15337.093                     having a wedding        adult   
9      0     23108.150  purchase of the house for my family        adult   
10     0     18230.959                      buy real estate        adult   
11     0     12331.077  

In [83]:
df['total_income'].sort_values()

14398      3306.762
12846      3392.845
15963      3418.824
1581       3471.216
14095      3503.298
            ...    
16953    273809.483
20526    274402.943
9068     276204.162
19340    352136.354
12257    362496.645
Name: total_income, Length: 21230, dtype: float64

In [84]:
df['total_income'].mean()

26489.032407842675

In [85]:
df['total_income'].mode()

0    22166.627
dtype: float64

In [86]:
df['total_income'].median()

22925.296

based off of these ranges I will do less than 20000 total income as 'low income'
20000 to 100000 as 'middle income' and above 100000 as 'high income'.

In [87]:
def income_level(total_income):
    if total_income < 20000:
        return 'low income'
    if 20000 < total_income < 100000:
        return 'middle income'
    return 'high income'

In [88]:
df['income_level']= df['total_income'].apply(income_level)
df.head(20)

index  children  days_employed  dob_years            education  \
0       0         1           8437         42    bachelor's degree   
1       1         1           4024         36  secondary education   
2       2         0           5623         33  secondary education   
3       3         3           4124         32  secondary education   
4       4         0         340266         53  secondary education   
5       5         0            926         27    bachelor's degree   
6       6         0           2879         43    bachelor's degree   
7       7         0            152         50  secondary education   
8       8         2           6929         35    bachelor's degree   
9       9         0           2188         41  secondary education   
10     10         2           4171         36    bachelor's degree   
11     11         0            792         40  secondary education   
12     12         0         365269         65  secondary education   
13     13         0           1846         54         some college   
14     14         0           1844         56    bachelor's degree   
15     15         1            972         26  secondary education   
16     16         0           1719         35  secondary education   
17     17         0           2369         33    bachelor's degree   
18     18         0         400281         53  secondary education   
19     19         0          10038         48  secondary education   

    education_id      family_status  family_status_id gender income_type  \
0              0            married                 0      F    employee   
1              1            married                 0      F    employee   
2              1            married                 0      M    employee   
3              1            married                 0      M    employee   
4              1  civil partnership                 1      F     retiree   
5              0  civil partnership                 1      M    business   
6              0            married                 0      F    business   
7              1            married                 0      M    employee   
8              0  civil partnership                 1      F    employee   
9              1            married                 0      M    employee   
10             0            married                 0      M    business   
11             1            married                 0      F    employee   
12             1  civil partnership                 1      M     retiree   
13             2            married                 0      F    employee   
14             0  civil partnership                 1      F    business   
15             1            married                 0      F    employee   
16             1            married                 0      F    employee   
17             0  civil partnership                 1      M    employee   
18             1    widow / widower                 2      F     retiree   
19             1           divorced                 3      F    employee   

    debt  total_income                              purpose    age_group  \
0      0     40620.102                purchase of the house        adult   
1      0     17932.802                         car purchase        adult   
2      0     23341.752                purchase of the house        adult   
3      0     42820.568              supplementary education        adult   
4      0     25378.572                    to have a wedding  older adult   
5      0     40922.170                purchase of the house  young adult   
6      0     38484.156                 housing transactions        adult   
7      0     21731.829                            education        adult   
8      0     15337.093                     having a wedding        adult   
9      0     23108.150  purchase of the house for my family        adult   
10     0     18230.959                      buy real estate        adult   
11     0     12331.077  

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Good job.
</div>

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [89]:

100*(df_cat.groupby('children')['debt'].mean()).sort_values()



children
5    0.000000
0    7.545824
3    8.231707
1    9.202838
2    9.514468
4    9.756098
Name: debt, dtype: float64

**Conclusion**

On average there is very little trend for number of children and the chance the borrower defaulting. In fact according to our data families with 5 children have never defaulted on their loans. We have a very small sample for such instances but I do not see evidence of a higher likelyhood of default based on children. There is a higher likelyhood of default if a person has children vs if they do not. About 2 percent difference.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

You are right.
</div>

**Is there a correlation between family status and paying back on time?**

In [90]:

100*(df_cat.groupby('family_status')['debt'].mean()).sort_values()




family_status
widow / widower      6.553911
divorced             7.124682
married              7.557521
civil partnership    9.314202
unmarried            9.784173
Name: debt, dtype: float64

**Conclusion**

 Those who are widowed, divorced or married are about 2% less likely to default vs. those in a civil partnership or those who are unmarried. 

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

<s>Why? 9.78 / 7.56 -> 30% higher!
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correct.
</div>

**Is there a correlation between income level and paying back on time?**

In [91]:
100*(df.groupby('income_level')['debt'].mean()).sort_values()





income_level
high income      6.060606
middle income    8.022774
low income       8.321283
Name: debt, dtype: float64

**Conclusion**

Lower and middle incomes are 2 percent more likely to default on thier loans. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correct.
</div>

**How does credit purpose affect the default rate?**

In [92]:
100*(df.groupby('purpose_category')['debt'].mean()).sort_values()

purpose_category
housing            7.201029
real estate        7.539282
wedding            7.872988
car transaction    9.323626
education          9.348534
Name: debt, dtype: float64

**Conclusion**
Car transactions and education have 2% more risk of default vs. wedding, housing, and real estate

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Great.
</div>

# General Conclusion 

My analysis is complete. I have resolved many issues with the data, including replacing missing values (populated with optimized values based on each entry), removed implicit duplicates, fixed the data types and unified mix cases. I also built multiple categories to visualize the data. Through my analysis the bank will be able to develop a credit score for their borrowers. 

I was able to conclude that there is, on average, a 2% trend towards default when a borrower is: young to middle age, low to middle income, and borrowing for education or a car. These should be useful when considering the credit score.

I am grateful for the opportunity to handle these data. Thank you very much for your time reviewing my work. I am looking forward to your feedback.


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Good overall analysis!
</div>

<div class="alert alert-block alert-warning">
<b>Overall reviewer's comment</b> <a class="tocSkip"></a>

Johannah, thank you for sending your project. You've done a really good job on it! <br>
    
There are a lot of correct conclusions and nice code. <br>
    
However, there are some issues. I wrote comments. Please elaborate them.
    
You have very good analytic skills, keep up the good work! I will be waiting for your corrected project.
</div>

<div class="alert alert-block alert-success">
<b>Overall reviewer's comment v2</b> <a class="tocSkip"></a>

Johannah, I really appreciate the corrections you sent in! Thanks for taking the time to do so. <br>
    Now your project is a true "A". Congratulations! <br>
    Your project has passed code review. Keep up the good work!
</div>